compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/PointerMatcher.kt

```diff
@@ -0,0 +1,147 @@
+
+/**
+ * [PointerMatcher] represents a single condition or a set of conditions which a [PointerEvent] has to match
+ * in order to count as an appropriate event for a gesture.
+ *
+ *  Supported matchers:
+ * - [mouse] - will match an event with [PointerType.Mouse] with a required [PointerButton]
+ * - [touch] - will match any event with [PointerType.Touch]
+ * - [stylus] - will match an event with [PointerType.Stylus]. And optional [PointerButton] can be specified.
+ * - [eraser] - will match any event with [PointerType.Eraser]
+ * - [pointer] - takes in [PointerType] and optional [PointerButton]
+ *
+ * Their combination is supported using plus operator:
+ * ```
+ * mouse(PointerButton.Primary) + touch + stylus + eraser
+ * ```
+ * See [Primary].
+ */
+@ExperimentalFoundationApi
+@OptIn(ExperimentalComposeUiApi::class)
+interface PointerMatcher {
+
+    @ExperimentalFoundationApi
+    fun matches(event: PointerEvent): Boolean
+
+    @ExperimentalFoundationApi
+    operator fun plus(pointerMatcher: PointerMatcher): PointerMatcher {
+        return if (this is CombinedPointerMatcher) {
+            this.sources.add(pointerMatcher)
+            this
+        } else if (pointerMatcher is CombinedPointerMatcher) {
+            pointerMatcher.sources.add(this)
+            pointerMatcher
+        } else {
+            CombinedPointerMatcher(mutableListOf(this, pointerMatcher))
+        }
+    }
+
+    @Suppress("MemberVisibilityCanBePrivate")
+    companion object {
+        @ExperimentalFoundationApi
+        fun pointer(
+            pointerType: PointerType,
+            button: PointerButton? = null
+        ): PointerMatcher = object : PointerTypeAndButtonMatcher {
+            override val pointerType = pointerType
+            override val button = button
+        }
+
+        @ExperimentalFoundationApi
+        fun mouse(button: PointerButton): PointerMatcher = MousePointerMatcher(button)
+
+        @ExperimentalFoundationApi
+        fun stylus(button: PointerButton? = null): PointerMatcher = StylusPointerMatcher(button)
+
+        @ExperimentalFoundationApi
+        val stylus: PointerMatcher = StylusPointerMatcher.Companion
+
+        @ExperimentalFoundationApi
+        val touch: PointerMatcher = TouchPointerMatcher
+
+        @ExperimentalFoundationApi
+        val eraser: PointerMatcher = EraserPointerMatcher
+
+        private interface PointerTypeMatcher : PointerMatcher {
+            val pointerType: PointerType
+
+            override fun matches(event: PointerEvent): Boolean {
+                return event.changes.fastAll { it.type == pointerType }
+            }
+        }
+
+        private interface PointerTypeAndButtonMatcher : PointerTypeMatcher {
+            val button: PointerButton?
+
+            override fun matches(event: PointerEvent): Boolean {
+                return event.changes.fastAll { it.type == pointerType } &&
+                    if (button != null) event.button == button else true
```

Comment:
If someone wants to differentiate between a stylus press and stylus button press, then I think they need to just add 2 onClick modifiers in correct order:
```kotlin
onClick(matcher = PointerMatcher.stylus) {}
.onClick(matcher=PointerMatcher.stylus(PointerButton.Primary)) {}
```

If we had `event.button == button` instead of `if (button != null) event.button == button else true` then it means that button is required to be null. So we won't handle a simple stylus press with button pressed in cases when it's not important whether anything is pressed or not. Imagine taking a stylus away from the screen, then moving it to the screen and pressing on it.

Reply:
> So we won't handle a simple stylus press with button pressed

But isn't the `button` is a change of the state? I.e. even if press with a stylus, we have `button` as null. If we want to handle the case when we need a press when the primary button is pressed, we need to add to matcher `buttons`, not `button`.

Here, if we write:
`onClick(matcher = PointerMatcher.stylus) {}
.onClick(matcher=PointerMatcher.stylus(PointerButton.Primary)) {}`, we will have two callbacks called when we press the primary button on the stylus, because in the first `onClick` all presses passes the filter.

Reply:
I see your point. 

> But isn't the `button` is a change of the state?

yes, it's a change of the state. So my example was incorrect. 

I agree, that `event.button == button` makes more sense now. I'll commit the change with your suggestion.

So when `matcher=PointerMatcher.stylus`, it   would mean that it's required that no button should be pressed on the stylus. it's equivalent to a click using touch.